### PR TITLE
GH-221 Participant Behaviour & Actions

### PR DESCRIPTION
**Pull request by fericsepi**
### Introducing:
 - add helper methods to exercise so its convenient to fetch just the phishing emails
 - add helper method to participant so its easy to gather actions specific to a given email and judge user behaviour based on those actions

### Example Output
Method `Participant.phishing_email_behaviour_and_actions` requires an `email_id` and return a `tuple` of `(behaviour string, actions list)`

image

### CODE REVIEW

File: participant/models.py [hunk 1]
```diff
@@ -74,6 +76,31 @@ def actions(self):
...
+        """
+        Aggregates actions related to a phishing email and determines behaviour.
+
+        :param email_id: UUID - id of an email
+        :return: (STRING, LIST) - returns a tuple of participant behaviour and actions
+        """
+
+        def participant_behaviour(actions):
+            pb = ParticipantBehaviour.NEUTRAL
+            if [a for a in actions if a.get('action_type') in POSITIVE_ACTIONS]:
```
Comment (DanielGeoffreyHodgkins) [inline participant/models.py]: if this is a list comprehension won't this only work on action[0]

Comment (fericsepi) [reply participant/models.py]: list comprehensions return a list regardless so this should work fine

Comment (DanielGeoffreyHodgkins) [reply participant/models.py]: can actions contain both positive and negative, and if so, overwriting `pb` with negative is always what we want? If so then 👍 just making sure!

Comment (fericsepi) [reply participant/models.py]: spot on - negatives overrule positives

File: participant/models.py [hunk 2]
```diff
@@ -74,6 +76,31 @@ def actions(self):
...
+                pb = ParticipantBehaviour.POSITIVE
+            if [a for a in actions if a.get('action_type') in NEGATIVE_ACTIONS]:
+                pb = ParticipantBehaviour.NEGATIVE
+            return pb
+
+        actions = []
+
+        if email_id in self.exercise.phishing_email_ids:
+            for action_id, action_details in self.actions.items():
+                if email_id == action_details.get('email_id'):
```
Comment (asgharsir) [inline participant/models.py]: Reference to the issue https://github.com/cybsafe/phishtray/issues/221. I wonder if it is filtering the "top actions only".

Comment (fericsepi) [reply participant/models.py]: good point - the first action that the user made should be the first item in the list
we went through the specs with Ben yesterday and found that we won't need to separately send the first item as the FE guys can just take the first item from the array

note - if we decide to send the first item separately then we can look at introducing some logic that looks at the `timedeltas` of each item and extract it that way

Comment (asgharsir) [reply participant/models.py]: This is why I made a separate property only for the first ones. Anyways with this resolved no point in reviewing that PR now.

**Pull request merged**